### ISSUE TITLE
Unable to compile/test with race detector enabled

### ISSUE THREAD

**Issue by princjef** [microsoft/go#904]
Apologies in advance if this is the wrong place to post this issue. This appears to be the most relevant place for issues since https://github.com/microsoft/go-images does not allow them.

I'm working with the MCR docker images for golang and have encountered an apparent gap in support for compilation and testing with the `-race` flag enabled. It's unclear from both that repo and this one whether the missing support is an intentional omission or an oversight. This issue hampers our ability to perform automated verification of data races in our build workflows with docker.

**Images tried**

- mcr.microsoft.com/oss/go/microsoft/golang:1.20
- mcr.microsoft.com/oss/go/microsoft/golang:1.20-fips-bullseye

**Failing commands**

- go test -race ./...
- go build -race .

**Minimal repro**

Using the MCR build, a hello world program compiles successfully without `-race`, but fails in linking when passing `-race`:

```
> docker run --rm -it mcr.microsoft.com/oss/go/microsoft/golang:1.20 bash

root@622f6e61f56e:/go# mkdir /app
root@622f6e61f56e:/go# cd /app
root@622f6e61f56e:/app# go mod init test
go: creating new go.mod: module test
root@622f6e61f56e:/app# echo '
> package main
>
> import "fmt"
>
> func main() {
>     fmt.Println("Hello world")
> }
> ' > main.go
root@622f6e61f56e:/app# go build .
root@622f6e61f56e:/app# go build -race .
# test
runtime.(*mspan).setUserArenaChunkToFault: relocation target __tsan_free not defined
runtime.mallocgc: relocation target __tsan_malloc not defined
runtime.runfinq: relocation target __tsan_finalizer_goroutine not defined
runtime.(*sweepLocked).sweep: relocation target __tsan_free not defined
runtime.goexit1: relocation target __tsan_go_end not defined
runtime.gfget: relocation target __tsan_malloc not defined
runtime.(*p).init: relocation target __tsan_proc_create not defined
runtime.(*p).destroy: relocation target __tsan_go_end not defined
runtime.(*p).destroy: relocation target __tsan_proc_destroy not defined
runtime.RaceDisable: relocation target __tsan_go_ignore_sync_begin not defined
runtime.RaceEnable: relocation target __tsan_go_ignore_sync_end not defined
runtime.raceinit: relocation target __tsan_init not defined
runtime.raceinit: relocation target __tsan_map_shadow not defined
runtime.racefini: relocation target __tsan_fini not defined
runtime.racemapshadow: relocation target __tsan_map_shadow not defined
runtime.racegostart: relocation target __tsan_go_start not defined
runtime.raceacquireg: relocation target __tsan_acquire not defined
runtime.raceacquirectx: relocation target __tsan_acquire not defined
runtime.racereleaseg: relocation target __tsan_release not defined
runtime.racereleaseacquireg: relocation target __tsan_release_acquire not defined
runtime.racereleasemergeg: relocation target __tsan_release_merge not defined
/usr/local/go/pkg/tool/linux_amd64/link: too many errors
root@622f6e61f56e:/app#
```

However, with the official golang distribution it works:

```
> docker run --rm -it golang:1.20 bash

root@4896b6f6916c:/go# mkdir /app
root@4896b6f6916c:/go# cd /app
root@4896b6f6916c:/app# go mod init test
go: creating new go.mod: module test
root@4896b6f6916c:/app# echo '
> package main
>
> import "fmt"
>
> func main() {
>     fmt.Println("Hello world")
> }
> ' > main.go
root@4896b6f6916c:/app# go build .
root@4896b6f6916c:/app# go build -race .
root@4896b6f6916c:/app#
```

**Comment by dagood** [comment]
This is the right place--https://github.com/microsoft/go-images/blob/microsoft/main/SUPPORT.md mentions it, but maybe we should also say that in the readme itself, or use an issue template. Thanks for making your way here to report the issue. 😄 

From what I can see, these errors can show up when cgo is not enabled, e.g. after cross-compiling (https://github.com/golang/go/issues/42382). However, this issue does repro for me on a linux-amd64 host and linux-amd64 container with `CGO_ENABLED=true` according to `go env`, so that isn't what's going on here.

It looks like go 1.20 included a change to how the race detector is built/packaged that broke our packing scripts, causing us to miss the syso files. Go 1.19 images work fine with those steps.

I'll look into this. It will likely be a quick fix to some pattern matching and we can run a new microsoft/go release (1.20.3-2) so you can use it.

(I believe there's an upstream issue for Go to provide something like `go tool distro` so we don't need to maintain our own distribution scripts, but I can't find it at the moment.)

**Comment by dagood** [comment]
`-distpack` is what I had in mind, and it was merged into upstream `master` last week:
https://go-review.googlesource.com/c/go/+/470676.

The original x/build/cmd/release command was deleted in https://github.com/golang/build/commit/97ad0ce4dd678045d4f339e43db7581c4c364546, and it looks like the code we need to emulate now lives around here: https://github.com/golang/build/blob/208db0ba5f4e4fdfdc462e9c0af3f0e0999f6d45/internal/task/buildrelease.go#L225-L285

**Comment by dagood** [comment]
This is now fixed in 1.20.3-2. (Fresh mcr.microsoft.com/oss/go/microsoft/golang:1.20 etc.) Thanks again for the report!

**Comment by princjef** [comment]
Works like a charm 💯

Thanks @dagood!
